### PR TITLE
do not rely on slicing to avoid panics

### PR DESCRIPTION
**Pull request by matteosister**
solves #17

### CODE REVIEW

File: positional/src/error.rs [hunk 1]
```diff
@@ -8,6 +8,12 @@ pub type PositionalResult<T> = Result<T, PositionalError>;
 pub enum PositionalError {
     #[error("Unable to parse the file")]
     UnparsableFile,
+
     #[error("Unable to find a matcher for row `{0}`")]
     UnmatchedVariant(String),
+
+    #[error(
+        "The row passed is too small to work with the fields definition. The row needs to have at least {0} unicode chars. Passed row is: `{1}`"
```
Comment (denmuprima) [inline positional/src/error.rs]: it would be nice to indicate the length of the String in the error message.
The syntax for that is
> "[...] Passed row is: `{1}` with length {}", .1.len()

Note that the tests would need to be adjusted since they compare the error message to the expected text.

File: positional/src/error.rs [hunk 2]
```diff
@@ -8,6 +8,12 @@ pub type PositionalResult<T> = Result<T, PositionalError>;
...
+    )]
+    RowSizeError(usize, String),
```
Comment (denmuprima) [inline positional/src/error.rs:18]: This and other things in this PR are major version changes, so we need to bump the major version.

File: positional/tests/parse_row.rs [hunk 3]
```diff
@@ -35,3 +35,27 @@ fn ser_de() {
...
+fn string_smaller_than_field_definition() {
+    let row_content = "1    ---10";
+    let row = <Data as FromPositionalRow>::from_positional_row(row_content);
+    assert_eq!(
+        row.err().map(|e| e.to_string()),
+        Some(
+            format!("The row passed is too small to work with the fields definition. The row needs to have at least 30 unicode chars. Passed row is: `{}`", row_content)
+        )
+    );
+}
```
Comment (denmuprima) [inline positional/tests/parse_row.rs:60]: I would've liked to see a test with a type with no fields and an empty String; however, that's rejected by an error at compile time from the derive Macro. The error message isn't the best, but it's probably not a use case for the library anyway to handle types like that, so that's probably not a big problem.

File: positional/src/parsed_field.rs [hunk 4]
```diff
@@ -22,14 +22,20 @@ impl<'s> PositionalParsedField<'s> {
...
+        // function
+        let raw_value = self
+            .row
+            .chars()
+            .skip(self.offset)
+            .take(self.size)
+            .collect::<String>();
         if self.left_aligned {
-            raw_value.trim_end_matches(self.filler)
+            raw_value.trim_end_matches(self.filler).to_string()
```
Comment (denmuprima) [inline positional/src/parsed_field.rs:36]: We talked about the cloning offline. I don't think there's a simple way to avoid it if we want to keep the Unicode point handling. It's not perfect (doesn't handle grapheme clusters), but I prefer this over the old version. 

Your call as to whether you want to keep the simple byte-based indexing (though in that point, a disclaimer about not handling unicode data would be warranted in the API docs).

File: positional/src/parsed_field.rs [hunk 5]
```diff
@@ -22,14 +22,20 @@ impl<'s> PositionalParsedField<'s> {
...
-        let slice_start = self.offset;
-        let slice_end = self.offset + self.size;
-        let raw_value = &self.row[slice_start..slice_end];
+    pub fn to_value(&self) -> String {
+        // we don't take into consideration the empty string (or generally an out of bound string)
+        // because we are already checking for string correctness in the main `from_positional_row`
+        // function
+        let raw_value = self
+            .row
+            .chars()
```
Comment (denmuprima) [inline positional/src/parsed_field.rs:31]: Technically speaking, this is more than what's required for the PR since the API change itself already handles strings of insufficient length. 

This is, however, a big improvement over the existing code since it doesn't rely on byte boundaries, but actually works on unicode points. It doesn't handle localization completely because it doesn't iterate over grapheme clusters, but that should be outside of the scope of this PR and would require introducing another dependency, such as https://docs.rs/unicode-segmentation/latest/unicode_segmentation/. Note that working on Unicode points isn't perfect because a "character" can consist of multiple such points, which would be in one grapheme cluster.

File: positional/tests/parse_row.rs [hunk 6]
```diff
@@ -35,3 +35,27 @@ fn ser_de() {
         FromPositionalRow::from_positional_row(&row).expect("error converting from positional row");
     assert_eq!(original_data, data);
 }
+
+#[test]
+fn empty_string() {
+    let row = <Data as FromPositionalRow>::from_positional_row("");
```
Comment (denmuprima) [inline positional/tests/parse_row.rs:41]: You don't need to qualify the trait here, you can just call `Data::from_positional_row()`.

**Comment by MaeIsBad** [comment]
oops, I implemented the same fix(https://github.com/primait/positional.rs/pull/19) but I think your version is way better

One suggestion: do you think it makes sense to replace the matchers in examples with non panicking versions(https://github.com/primait/positional.rs/pull/19/commits/0c5a647180eed0f91d1c653d25c461c3ba681e92)?

**Comment by matteosister** [comment]
@MaeIsBad well the two PRs are basically the same, so mine is not WAY better at all!
And yes, non-panicking matches are better!

File: positional/src/parsed_field.rs (hunk 5 above)
Comment (matteosister) [reply positional/src/parsed_field.rs:31]: I think that we could add that library as a dependency behind a feature!

Comment (MaeIsBad) [reply positional/src/parsed_field.rs:31]: > This is, however, a big improvement over the existing code since it doesn't rely on byte boundaries, but actually works on unicode points

I think this is wrong, since string indexing in rust already works on unicode points

Comment (denmuprima) [reply positional/src/parsed_field.rs:31]: I tested this on playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=63b9782f5832a970b80c0d724bb39787...

EDIT 1: Oh curses, you're right. I forgot to add an `=` in my range. I'm literally querying an empty range.
Thanks for pointing this out!

EDIT 2: nope, the range is fine as-intended. I got confused because we were discussing something on Slack that contained an inclusive range, and I used an exclusive one here. Commented again below.

Comment (denmuprima) [reply positional/src/parsed_field.rs:31]: > I think that we could add that library as a dependency behind a feature!

I'm not sure. Gating additional content behind feature gates makes sense, I'm not sure the same can be said about changing existing behaviour. At any rate, it's out of scope for this PR, so let's discuss that elsewhere :)

Comment (denmuprima) [reply positional/src/parsed_field.rs:31]: Hang on, I got confused earlier. My original comment was right, and the playground example is fine. 
It compares `s.get(0..1)` with `s.chars().take(1)`. In both cases, you'd expect to get one character, but the get version returns None because it's not on a valid UTF-8 boundary. I stand by my original comment.

Comment (MaeIsBad) [reply positional/src/parsed_field.rs:31]: you're right, I was really sure about this... Oh well, thanks for teaching me something

Comment (matteosister) [reply positional/src/parsed_field.rs:31]: so it's better to use the chars iterator right?

Comment (MaeIsBad) [reply positional/src/parsed_field.rs:31]: Yes

File: positional/src/parsed_field.rs (hunk 4 above)
Comment (MaeIsBad) [reply positional/src/parsed_field.rs:36]: Would something like this work?
```rs
    /// output a string representation of the parsed value
    /// trimming is done by the library based on the declared positional row configurations
    pub fn to_value(&self) -> &str {
        // we don't take into consideration the empty string (or generally an out of bound string)
        // because we are already checking for string correctness in the main `from_positional_row`
        // function
        let mut indices_iter = self
            .row
            .char_indices()
            .map(|(i, _c)| i)
            .skip(self.offset)
            .take(self.size);

        let start = indices_iter.next();
        let end = indices_iter.last();

        let raw_value = match (start, end) {
            (Some(start), Some(end)) => self.row.get(start..=end).unwrap(),
            (Some(start), None) => self.row.get(start..).unwrap(),
            (None, None) => self.row.get(..).unwrap(),
            // If the char_indices iterator returned None there is no way for it to return Some
            (None, Some(_)) => unreachable!(),
        };

        if self.left_aligned {
            raw_value.trim_end_matches(self.filler)
        } else {
            raw_value.trim_start_matches(self.filler)
        }
    }
```
there is probably a few ways to make this code prettier, this is just a poc

File: positional/src/error.rs (hunk 2 above)
Comment (denmuprima) [reply positional/src/error.rs:18]: Discussed offline; we have more changes coming up, so we'll bump it in a later PR.

File: positional/src/parsed_field.rs (hunk 4 above)
Comment (MaeIsBad) [reply positional/src/parsed_field.rs:36]: Actually we could just do 
```rs
    /// output a string representation of the parsed value
    /// trimming is done by the library based on the declared positional row configurations
    pub fn to_value(&self) -> &str {
        let mut indices_iter = self
            .row
            .char_indices()
            .map(|(i, _c)| i)
            .skip(self.offset)
            .take(self.size);

        // we don't take into consideration the empty string (or generally an out of bound string)
        // because we are already checking for string correctness in the main `from_positional_row`
        // function
        let start = indices_iter.next().unwrap();
        let end = indices_iter.last().unwrap();

        // This could be a get_unchecked, but the compiler should be able to optimize it on it's own
        let raw_value = self.row.get(start..=end).unwrap();

        if self.left_aligned {
            raw_value.trim_end_matches(self.filler)
        } else {
            raw_value.trim_start_matches(self.filler)
        }
    }
```
because of the oob checks

Comment (matteosister) [reply positional/src/parsed_field.rs:36]: I was talking with @denmuprima yesterday about relying on indices to avoid cloning. We could go with this. I'm ok. My only concern is that the actual code is more terse and clear, and probably the impact is really negligible. Up to you. What should we do?

Comment (MaeIsBad) [reply positional/src/parsed_field.rs:36]: I don't want to get in your way on this PR too much, it's up to you. I do quite like doing optimizations like these, especially in libraries, but I don't think the performance benefits will end up actually ever mattering.

Comment (matteosister) [reply positional/src/parsed_field.rs:36]: I just remembered that I also created a bench suite for this library, gonna try this!

Comment (denmuprima) [reply positional/src/parsed_field.rs:36]: https://doc.rust-lang.org/std/primitive.str.html#method.char_indices
> As a string slice consists of valid UTF-8, we can iterate through a string slice by [char](https://doc.rust-lang.org/std/primitive.char.html). This method returns an iterator of both these [char](https://doc.rust-lang.org/std/primitive.char.html)s, as well as their byte positions.
This would generally work, yes.

Good suggestion! I don't think this approach works as-is. 
I thought about using that method yesterday, but I figured it might complicate the code too much; however, there's some simplifications we could do:
That (None, None) case should be unreachable, no? It should be if we disallow "empty" fields. Thoughts @matteosister? 
This is a decent idea to avoid the allocation through `collect`, for sure, thanks for bringing it up!
I see two problems with this version: I think you need to `take` one more element here; the original only collected the characters that belong to each column/field, but we need the byte boundary where the next element starts. 
Moreover, it won't properly recognize the end of the current string. `take` will only give you as many elements as are available. If you have a one character String, for instance, start and end would be the same. The easiest way to fix this is probably by changing the interface a little; we know when we're handling the last element in `from_positional_row` and could pass that information here.
I adjusted my playground a little: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=64c3a2e86471d05db90d63f0f659e7fd. So yes, this would work.

Overall, I don't know if it's worth it to avoid the copy, though. I guess we should decide whether we want to support arbitrary unicode characters or keep acting on individual unicode points (as in the current version of the PR); if the latter, let's go for char_indices :) 


Side note: also discussed offline, we'll add a test to handle limited unicode strings (deserializing into a type with a single field and a single letter like "ë").

File: positional/src/error.rs [hunk 7]
```diff
@@ -8,6 +8,12 @@ pub type PositionalResult<T> = Result<T, PositionalError>;
 pub enum PositionalError {
     #[error("Unable to parse the file")]
     UnparsableFile,
+
     #[error("Unable to find a matcher for row `{0}`")]
     UnmatchedVariant(String),
+
+    #[error(
+        "The row passed is too small to work with the fields definition. The row needs to have at least {0} unicode chars. Passed row is: `{1}` with length {}", .1.len()
```
Comment (denmuprima) [inline positional/src/error.rs]: I know wording at the end comes from my suggestion, but upon seeing it again, maybe replacing the entire thing with
> Given row `{1}` has length {}; expected length {0}", .1.len()
would sound better?

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### [Next]
 
+- support for unicode chars while parsing files.
```
Comment (denmuprima) [inline CHANGELOG.md]: Limited support. 
https://doc.rust-lang.org/std/primitive.str.html#method.chars 
We're respecting UTF-8 scalar values, but aren't taking Unicode into consideration fully.

File: positional/src/parsed_field.rs (hunk 4 above)
Comment (denmuprima) [reply positional/src/parsed_field.rs:36]: Sorry; it took me a bit to write that comment, so I didn't see your second version! It simplifies the match, which is nice, but the problem with the .take() and the last element still apply.

Benching it is a good idea; however, I doubt anyone would use this in a performance-sensitive piece of code? I don't think the extra allocation is that big a deal. Regardless, we could move this to a separate PR for the perf improvement?

File: positional/tests/parse_row.rs [hunk 9]
```diff
@@ -35,3 +35,34 @@ fn ser_de() {
...
+            format!("The row passed is too small to work with the fields definition. The row needs to have at least 30 unicode chars. Passed row is: `{}` with length {}", row_content, row_content.len())
+        )
+    );
+}
+
+#[test]
+fn unicode_chars() {
+    let row = Data::from_positional_row("Noël ---10the address is this ")
+        .expect("error converting from positional row");
+    assert_eq!(Data::new("Noël", 10, "the address is this"), row);
```
Comment (denmuprima) [inline positional/tests/parse_row.rs:66]: Does this fail with the previous implementation?

Comment (MaeIsBad) [reply positional/tests/parse_row.rs:66]: It used to, I edited my comment. I did `start..end` instead of `start..=end`

Comment (denmuprima) [reply positional/tests/parse_row.rs:66]: Just want to make sure the test fails before this PR and works with the changes, so we're sure we're testing the right thing!

File: positional/src/parsed_field.rs (hunk 4 above)
Comment (matteosister) [reply positional/src/parsed_field.rs:36]: ok the results are:

with string slice:
```
deserialize 100.000 structs                                                                          
                        time:   [348.79 ms 349.96 ms 351.21 ms]
                        change: [+7.1842% +7.8299% +8.4109%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

with clone:
```
deserialize 100.000 structs                                                                          
                        time:   [352.23 ms 354.01 ms 355.71 ms]
                        change: [+8.3435% +9.0796% +9.7860%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

a really small difference. We are talking about 5 ms less to deserialize 100.000 structs from a row.
I think I'll keep the actual one! @MaeIsBad @denmuprima

File: positional/tests/parse_row.rs (hunk 9 above)
Comment (matteosister) [reply positional/tests/parse_row.rs:66]: yes! On master this fails:
`thread 'unicode_chars' panicked at 'error converting from positional row: ParseIntError { kind: InvalidDigit }', positional/tests/parse_row.rs:42:10`

It takes the first 5 digits, resulting in `Noël` without the ending space.
and for the next 5 as age, it gets ` ---1` with an initial space. That fails to be parsed as an i32.

**Comment by matteosister** [comment]
Everything should be fixed!

**Comment by matteosister** [comment]
also, we are still in the 0. phase for the library. So I think that we'll go with a 0.3 version.

**Comment by denmuprima** [comment]
> also, we are still in the 0. phase for the library. So I think that we'll go with a 0.3 version.

The current version is  0.2.3, so 0.3 is indeed the correct version. Cargo semver has stricter requirements for breaking changes than pure semver for 0.x.y versions since you need to bump x.

File: positional/tests/parse_row.rs (hunk 9 above)
Comment (denmuprima) [reply positional/tests/parse_row.rs:66]: Awesome, thank you!

File: positional/src/parsed_field.rs (hunk 4 above)
Comment (denmuprima) [reply positional/src/parsed_field.rs:36]: Thanks for checking this out!
I wonder if we should have benchmarks with larger text because I do expect a larger difference on larger input, but we can tackle that another time.

File: positional/benches/enum.rs [hunk 10]
```diff
@@ -85,7 +85,7 @@ fn deserialize_enum(c: &mut Criterion) {
 criterion_group! {
     name = benches;
     // This can be any expression that returns a `Criterion` object.
-    config = Criterion::default().sample_size(60);
+    config = Criterion::default().sample_size(30);
```
Comment (denmuprima) [inline positional/benches/enum.rs:88]: I didn't expect to see changes for the benchmarks in this PR, but since it's such a small change, I don't mind.
Why are we reducing the sample size, though?

File: positional/src/error.rs [hunk 11]
```diff
@@ -13,7 +13,7 @@ pub enum PositionalError {
     UnmatchedVariant(String),
 
     #[error(
-        "The row passed is too small to work with the fields definition. The row needs to have at least {0} unicode chars. Passed row is: `{1}` with length {}", .1.len()
+        "Given row `{1}` has length {}; expected length: {0}", .1.len()
```
Comment (denmuprima) [inline positional/src/error.rs:16]: Thanks for adjusting this! I do think this looks much cleaner :)

**Review by denmuprima — DISMISSED** [review]
Thanks for iterating! Very nice improvement over the state of the lib before the changes.
LGTM!

File: positional/benches/enum.rs (hunk 10 above)
Comment (matteosister) [reply positional/benches/enum.rs:88]: because criterion needs a sample size small enough to be completed in 3 seconds. My PC is pretty old and bloated, so maybe on better PCs this is not needed. Anyway it's just useful for predicting the time needed to complete the bench

**Comment by matteosister** [comment]
@denmuprima the best one is Elm. That picks automatically the correct version based on the API changes.

**Review by MaeIsBad — APPROVED** [review]
LGTM
